File: Iterator/src/iterator2/FixedArrayIterator.java
package iterator2;

/**
 * @author Tudor
 *
 */
public class FixedArrayIterator implements Iterator {

	FixedArray array;
	int i = -1;
	
	FixedArrayIterator(FixedArray a){
		this.array = a;
	}
	
	@Override
	public Object CurrentItem() {
		if(i!=-1 && i!=array.size)
			return array.data[i];
		return null;
	}

	@Override
	public void First() {
		if(array.nrItems > 0)
			i = 0;
	}

	@Override
	public void Next() {
		if(i < array.nrItems)
			i++;
	}

	@Override
	public boolean isDone() {
		if(i == array.nrItems)
			return false;
		return true;
	}

}

File: Iterator3/Iterator3/FixedArrayIterator.cpp
#include "FixedArrayIterator.h"

template<class Item>
FixedArrayIterator<Item>::FixedArrayIterator(FixedSizeArray<Item>* aArray)
	:_array(aArray), _current(0)
{
	;//Nothing
}

template <class Item>
Item FixedArrayIterator<Item>::CurrentItem () const 
{
	if (IsDone()) {
		/*throw IteratorOutOfBounds*/;
	}
	return _array->Get(_current);
}

template<class Item>
void FixedArrayIterator<Item>::First() 
{
	_current = 0;
}

template<class Item>
void FixedArrayIterator<Item>::Next()
{
	_current++;
}

template<class Item>
bool FixedArrayIterator<Item>::IsDone() const
{
	return _current >= _array->Count();
}